Add a function to check remote HTTP/2 setting MAX_CONCURRENT_STREAMS

zuiderkwast opened the pull request:
This is to be used primarily by a client before initiating a new stream
so that it can respect the setting from the server.

I've tested in in a draft patch to gun. Will post a gun PR soon.

essen commented:
Right I believe the missing spec and the todo are why it didn't move forward. The todo in particular should be done. Perhaps it's better to maintain a counter in the state than count streams manually, since we want only locally initiated streams.

zuiderkwast commented:
I did the todo and added a spec.

> Perhaps it's better to maintain a counter in the state than count streams manually, since we want only locally initiated streams.

The local number of streams is never more than the total number of streams, so I added counting only if the total number reaches the limit. Thus, this is O(1) in hopefully most cases but O(n) in the worst case.

Do you prefer storing a counter in the state? There are multiple places where streams are added to and removed from the state so it would add some complexity.

essen commented:
Probably fine like this.

zuiderkwast commented:
Ping. This one is ready too. (Sorry for spamming.)

essen commented on src/cow_http2_machine.erl:
```diff
@@ -1613,6 +1614,27 @@ get_stream_remote_state(StreamID, State=#http2_machine{mode=Mode,
+			Sum + 1;
+		(_, Sum) ->
+			Sum
+	end, 0, maps:keys(Streams)).
```
Please change this into a `maps:fold` so we can avoid the building of the list.

zuiderkwast replied:
I did that first, assuming that `maps:fold` is efficient, but after some [investigation](https://github.com/erlang/otp/blob/master/lib/stdlib/src/maps.erl#L403) I found that it uses a maps iterator, which builds nested 3-tuples of the whole map with keys and values. This is just like building a list, but a 3-tuple uses 5 words while a cons cell uses 2 words.

Thus, I used `maps:keys` since it seemed more efficient, but I can switch to fold if you prefer.

This example shows how the iterator is represented:

```Erlang
1> Map = maps:from_list([{X, X} || X <- lists:seq(1,100)]).
#{12 => 12,4 => 4,38 => 38,16 => 16,93 => 93,53 => 53,
  92 => 92,46 => 46,28 => 28,23 => 23,77 => 77,59 => 59,
  88 => 88,80 => 80,58 => 58,30 => 30,90 => 90,73 => 73,
  96 => 96,40 => 40,100 => 100,47 => 47,24 => 24,27 => 27,
  89 => 89,21 => 21,97 => 97,29 => 29,22 => 22,...}
2> Iter = maps:iterator(Map).
[0|
 #{12 => 12,4 => 4,38 => 38,16 => 16,93 => 93,53 => 53,
   92 => 92,46 => 46,28 => 28,23 => 23,77 => 77,59 => 59,
   88 => 88,80 => 80,58 => 58,30 => 30,90 => 90,73 => 73, 
   96 => 96,40 => 40,100 => 100,47 => 47,24 => 24,27 => 27,
   89 => 89,21 => 21,97 => 97,...}]
3> maps:next(Iter).
{12,12,
 {4,4,
  {38,38,
   {16,16,
    {93,93,{53,53,{92,92,{46,46,{28,28,{23,23,...}}}}}}}}}}
```

zuiderkwast replied:
I timed `maps:next(maps:iterator(Map))` vs `maps:keys(Map)` and found that the iterator is faster, though I don't really understand why. I tried to see [how the bif is implemented](https://github.com/erlang/otp/pull/1594/files#diff-d4a9baa2fc8e210acb0b463811eb82d4352007ba86e66481f90e85f09afe5516R2949), which I haven't digged very deeply into, but I'm convinced `maps:fold` is better than `maps:keys`. I'll change it. :-)

essen replied:
Likely because optimisations can be applied in that scenario whereas they can't when you do maps:keys. Either way this isn't a super hot code path and if it was it'd be better replaced with a counter in the state.

essen commented on src/cow_http2_machine.erl:
```diff
@@ -1613,6 +1614,27 @@ get_stream_remote_state(StreamID, State=#http2_machine{mode=Mode,
+		map_size(Streams) >= MaxConcurrentStreams andalso
+		count_local_streams(State) >= MaxConcurrentStreams.
+
+-spec count_local_streams(http2_machine()) -> non_neg_integer().
```
Please drop the spec since it's a local function, Dialyzer can figure it out just fine.

essen commented:
The function looks good so can be merged regardless of the Gun PR's status.

essen commented:
Merged, thanks!!
